### ISSUE TITLE
Add support for sharing files/secrets with multiple users using Ansible Vault (gpg2 support)

### ISSUE THREAD

**Issue by ffledgling** [ansible/ansible#17346]
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

vault
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = 
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

N/A
##### SUMMARY

Ansible Vault currently uses AES encryption with a password to share and store files. When the playbooks are shared and owned by a team, sharing a plaintext password is cumbersome, error prone and a possible security concern. It would be nice if vault added support for encrypting files and secrets with public keys for a set of people. Ideally either encrypting the AES shared password with everyone's public keys or directly encrypting the secrets with public keys. Essentially multiple people in a team/contributing/using a repository containing ansible-playbooks should be able to deploy playbooks correctly without having to share a common password/piece of knowledge necessary for the deploy.
##### STEPS TO REPRODUCE

- Try to use ansible-vault with multiple people without needing  a common password.
- Fail. :(

**Comment by bcoca** [comment]
I'm not sure this is possible to implement, I've done "shamir's shared secret" but that is using multiple users to encrypt a secret and requiring only a subset to decrypt. 

Theoretically you could have that subset be 1 user but I'm not sure how secure that is.

**Comment by amenonsen** [comment]
It's possible (and straightforward) to implement. You just tell gpg to encrypt something with a set of public keys, then anyone who has any of the corresponding private keys can decrypt it. Works very well. (This is the reason I implemented vault+gpg integration, which is blocked by the fact that python-gnupg needs some patches first, which the maintainer has not yet replied to.)

**Comment by bcoca** [comment]
I'll look into that, but I thought you ended up with N encrypted versions of the secret, 1 per public key used.

**Comment by bcoca** [comment]
@amenonsen thanks, looked into GPG multiple keys, what ends up happening is that they use a generated key to encrypt the data, then create encrypted copies of this key for each public key.

So this should 'already work' via the GPG 'vault script'.

**Comment by ffledgling** [comment]
@bcoca I have a bash/wrapper script around GPG that does this already, basically the same as @amenonsen described. The only problem is that I now have to write additional tooling to encrypt the vault password with my makeshift script and then decrypt that and give it to ansible before it can deploy playbooks, this is not ideal and it'd be nice to have 1st class support for GPG in vault. If you need a dev to look at it, I'll be happy to take a stab at a patch if you can point me to the right direction.

**Comment by bcoca** [comment]
@ffledgling I see 2 options here, using a vault script to provide the vault secret (i.e. https://github.com/ansible/ansible/blob/devel/contrib/vault/vault-keyring.py) or add an additional encryption format 'GPG'. I'm not sure which is the best solution. At this point I think this should shift into a proposa http://github.com/ansible/proposals .

cc @alikins @abadger

**Comment by chrrrles** [comment]
My vote would be for adding GnuPG as an additional encryption format.  Wrapper scripts (obviously) are an additional layer of complexity...  and it just feels like PGP should already be supported in vault.

**Comment by deterralba** [comment]
This is probably not a high priority task, but still it is interesting.

Something like what [blackbox](https://github.com/StackExchange/blackbox) does would be a nice feature: being able to remove the access for one key, instead of using `rekey` to change the password.

Sadly, I don't have enough knowledge in crypto to help here.

**Comment by Mehonoshin** [comment]
This feature looks very valuable. It is sad, that it's not implemented yet.

**Comment by bendem** [comment]
I'd recommend using something like [gopass](https://github.com/gopasspw/gopass) which does this already and writing a small [password-client](https://docs.ansible.com/ansible/latest/user_guide/vault.html#vault-password-client-scripts) that redirects the password request to it.
Let the development effort on ansible be free of cryptographic challenges. One tool per job, simpler tools, more options.

**Comment by ffledgling** [comment]
@bendem I disagree with this approach. Ansible already strayed into the territory of secrets management the moment it introduced the concept of vault. The problem now is not whether it should or shouldn't, but how well it supports real-world use-cases. If ansible plans to continue supporting storage and retrieval of secrets then the current ecosystem, it is sub-par in the current state and should be improved. If it does not, then it needs to discontinue support for ansible-vault and add some kind of generic hooks/framework/platform that can tie into other existing solutions such as keypass or hashicorp-vault. There is some of this already in place via password-client-scripts, but it is less than perfect and still leaves a lot of room for "roll-your-own", which as stated is less than ideal for something that is so common for any non-trivial ansible deployment.

**Comment by goetzk** [comment]
I too would be interested in this behaviour - being able to use gpg2 for secrets would make vault much more usable in a team context.

I've had a look at the proposals list but didn't see anything relevant - has someone entered it and I didn't see it or was a proposal never opened?

Related proposals:
https://github.com/ansible/proposals/issues/24
https://github.com/ansible/proposals/issues/135

I've also considered putting secrets in a private bitwarden instance but at this stage the integration and maintenance effort seems to outweigh the benefits.

https://github.com/mickaelperrin/ansible-bitwarden-lookup-plugin

**Comment by ffledgling** [comment]
@goetzk I never actually ended up writing a proposal for this because the setup I was working in switched away from Ansible to something else. I still have the scripts and idea lying around so if you end up writing one, I'd be review/help flesh out if you need any such assistance.

**Comment by sivel** [comment]
This should be handled using a vault password script. The script would use gpg to decrypt a file that contains the central vault password file, and return that value.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
